### ISSUE TITLE
Setting *warn-on-reflection* prevents symbol renaming

### ISSUE THREAD

**Issue by andreyorst** [clojure-emacs/clj-refactor.el#485]
## Expected behavior

Symbol renamed

## Actual behavior

```
java.lang.IllegalStateException: refactor-nrepl is unable to build an AST for myapp.core. tools.analyzer encountered the following problem: Can't set!: *warn-on-reflection* from non-binding thread
```

## Steps to reproduce the problem

I've created a simple clojure file with `(set! *warn-on-reflection* true)` on top and tried renaming the symbol.

### clj-refactor.el version information

clj-refactor 2.5.1, refactor-nrepl 2.5.1

### CIDER version information

;; CIDER 1.1.0snapshot, nREPL 0.8.3
;; Clojure 1.10.3, Java 11.0.10

### Leiningen or Boot version

Leiningen 2.9.3 on Java 11.0.10 OpenJDK 64-Bit Server VM

### Emacs version

GNU Emacs 28.0.50 (build 1, x86_64-pc-linux-gnu, GTK+ Version 3.24.13, cairo version 1.16.0) of 2021-03-17

### Operating system

Fedora 33

**Comment by andreyorst** [comment]
Also when simply launching CIDER repl with jack in I see this in the `*Messages*` buffer:

    error in process filter: Some namespaces are in a bad state: error "Can’t set!: *warn-on-reflection* from non-binding thread" in myapp.core

for every namespace where I set this var. When working with GraalVM this setting is very handy and I would like to keep it.

**Comment by vemv** [comment]
Although popular, `(set! *warn-on-reflection*)` is not guaranteed to work. No top-level `(set! ... )` is - personally I keep seeing projects tripping up on this nuance.

Furthermore `(set! *warn-on-reflection*)`, while is thread-local, is not local to your ns, or to your project. So setting it can change its value for anyone consuming your ns.

tldr it's not an ideal mechanism.

What can you use instead:

* [lein check](https://github.com/technomancy/leiningen/blob/bdb2967b30b6d199dd173e4ec163dc9d4ca76142/src/leiningen/check.clj) and grep for reflection warnings
* Same, but with https://github.com/athos/clj-check
* Use https://github.com/jonase/eastwood#reflection as a linter

We're using Eastwood as a reflection linter in most https://github.com/clojure-emacs projects as a hard check in CI and it's working nicely for us. It has advantages over a simple grep laid out in its documentation.

(I'm biased though because I authored it)

**Comment by vemv** [comment]
I've (unsuccessfully) tried to reproduce the issue here https://github.com/clojure-emacs/refactor-nrepl/pull/325

Feel free to hack it, you might be able to break the build with a sophisticated-enough example :)
